### PR TITLE
[USMON-805] Fix HTTP/2 frame remainder bug

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

This PR fix couple of issues we had in our current support for HTTP/2:
- A use case in the get_first_frame process was identified where it was not possible and may have created issues, especially in edge cases with large payloads.
- There was a small bug related to frame_state->remainder, which was not handled correctly.

### Motivation

Improves our accuracy for HTTP/2 in both plain and TLS support.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 1]
```diff
@@ -438,7 +438,7 @@ static __always_inline void tls_fix_header_frame(tls_dispatcher_arguments_t *inf
 
 static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info, frame_header_remainder_t *frame_state, http2_frame_t *current_frame, http2_telemetry_t *http2_tel) {
     // No state, try reading a frame.
-    if (frame_state == NULL) {
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:441]: modify the comment
open a jira ticket to investigate why do we have a frame state where both remainder and header_length are zero

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 2]
```diff
@@ -625,6 +619,10 @@ int uprobe__http2_tls_handle_first_frame(struct pt_regs *ctx) {
 
     // skip HTTP2 magic, if present
     tls_skip_preface(&dispatcher_args_copy);
+    if (dispatcher_args_copy.data_off == dispatcher_args_copy.data_end) {
+    // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:623]:
```suggestion
        // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 3]
```diff
@@ -488,28 +488,22 @@ static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info
...
-        }
-    }
-
     // We failed to read a frame, if we have a remainder trying to consume it and read the following frame.
     if (frame_state->remainder > 0) {
+         if (info->data_off + frame_state->remainder > info->data_end) {
+            frame_state->remainder -= info->data_end - info->data_off;
+            info->data_off = info->data_end;
+            return false;
+        }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:497]: document

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 4]
```diff
@@ -639,6 +633,10 @@ int socket__http2_handle_first_frame(struct __sk_buff *skb) {
 
     // skip HTTP2 magic, if present
     skip_preface(skb, &dispatcher_args_copy.skb_info);
+    if (dispatcher_args_copy.skb_info.data_off == dispatcher_args_copy.skb_info.data_end) {
+    // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:637]:
```suggestion
        // Abort early if we reached to the end of the frame (a.k.a having only the HTTP2 magic in the packet).
```

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h (hunk 1 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/http2/decoding-tls.h:441]: https://datadoghq.atlassian.net/browse/USMON-797

**Review by guyarb — APPROVED** [review]

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 5]
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
 	}
 }
 
+func (s *usmHTTP2Suite) TestRemainderTable() {
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:1382]: Let's document the purpose of the test

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 6]
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
...
+		{
+			name: "validate clean with remainder and header zero",
+			// The purpose of this test is to validate that we cannot handle reassembled tcp segments.
+			messageBuilder: func() [][]byte {
+				a := newFramer().
+					writeHeaders(t, 1, usmhttp2.HeadersFrameOptions{Headers: testHeaders()}).bytes()
+				b := newFramer().writeData(t, 1, true, []byte("test12345")).bytes()
+				message := append(a, b[11:]...)
+				return [][]byte{
+					// we split it in 10 bytes in order to split the payload itself.
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:1408]: rephrase the comment

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 7]
```diff
@@ -1379,6 +1379,80 @@ func (s *usmHTTP2Suite) TestDynamicTable() {
...
+				a := newFramer().
+					writeHeaders(t, 1, usmhttp2.HeadersFrameOptions{Headers: testHeaders()}).
+					writeData(t, 1, true, emptyBody).bytes()
+				return [][]byte{
+					// we split it in 10 bytes in order to split the payload itself.
+					a[:10],
+					a[10:],
+				}
+			},
+			mapSize: 1,
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:1428]: out of scope, but isn't that a bug? shouldn't we clean the remainder?

Comment (amitslavin) [reply pkg/network/usm/usm_http2_monitor_test.go:1430]: https://datadoghq.atlassian.net/browse/USMON-804

**Comment by amitslavin** [comment]
@guyarb, I have reverted the changes to the payload size as it was causing different issues. I will address it in a separate PR. I've also updated the description and title of the PR. Could you please review it again to validate the current changes?

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 8]
```diff
@@ -437,8 +437,8 @@ static __always_inline void tls_fix_header_frame(tls_dispatcher_arguments_t *inf
 }
 
 static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info, frame_header_remainder_t *frame_state, http2_frame_t *current_frame, http2_telemetry_t *http2_tel) {
-    // No state, try reading a frame.
-    if (frame_state == NULL) {
+    // Attempting to read the initial frame in the packet, or handling a state where there is no remainder and finishing reading the current frame.
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:441]: do we still need that change?
didn't we fix the case of having empty remainder in this PR?

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 9]
```diff
@@ -435,8 +435,8 @@ static __always_inline void fix_header_frame(struct __sk_buff *skb, skb_info_t *
 }
 
 static __always_inline bool get_first_frame(struct __sk_buff *skb, skb_info_t *skb_info, frame_header_remainder_t *frame_state, http2_frame_t *current_frame, http2_telemetry_t *http2_tel) {
-    // No state, try reading a frame.
-    if (frame_state == NULL) {
+    // Attempting to read the initial frame in the packet, or handling a state where there is no remainder and finishing reading the current frame.
+    if (frame_state == NULL || (frame_state->remainder == 0 && frame_state->header_length == 0)) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:439]: same as tls version

**Review by guyarb — COMMENTED** [review]
reviewed

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 10]
```diff
@@ -467,7 +467,7 @@ static __always_inline bool tls_get_first_frame(tls_dispatcher_arguments_t *info
     //  4. We failed reading any frame. Aborting.
 
     // Frame-header-remainder.
-    if (frame_state != NULL && frame_state->header_length == HTTP2_FRAME_HEADER_SIZE) {
+    if (frame_state == NULL) {
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:470]: revert

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h (hunk 8 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding-tls.h:441]: still relevant

**Review by guyarb — APPROVED** [review]

**Pull request merged**